localization/pose_instability_detector/src/pose_instability_detector.cpp

```diff
@@ -0,0 +1,185 @@
+  diff_angle_z_pub_ = this->create_publisher<Float64Stamped>("~/debug/diff_angle_z", 10);
+  diagnostics_pub_ = this->create_publisher<DiagnosticArray>("/diagnostics", 10);
+}
+
+void PoseInstabilityDetector::callback_odometry(Odometry::ConstSharedPtr odometry_msg_ptr)
+{
+  latest_odometry_ = *odometry_msg_ptr;
+}
+
+void PoseInstabilityDetector::callback_twist(
+  TwistWithCovarianceStamped::ConstSharedPtr twist_msg_ptr)
+{
+  twist_buffer_.push_back(*twist_msg_ptr);
+}
+
+void PoseInstabilityDetector::callback_timer()
+{
+  if (!set_first_odometry_) {
+    prev_odometry_ = latest_odometry_;
+    set_first_odometry_ = true;
+    return;
+  }
+
+  auto quat_to_rpy = [](const Quaternion & quat) {
+    tf2::Quaternion tf2_quat(quat.x, quat.y, quat.z, quat.w);
+    tf2::Matrix3x3 mat(tf2_quat);
+    double roll{};
+    double pitch{};
+    double yaw{};
+    mat.getRPY(roll, pitch, yaw);
+    return std::make_tuple(roll, pitch, yaw);
+  };
+
+  Pose pose = prev_odometry_.pose.pose;
+  rclcpp::Time prev_time = rclcpp::Time(prev_odometry_.header.stamp);
+  for (const TwistWithCovarianceStamped & twist_with_cov : twist_buffer_) {
+    const Twist twist = twist_with_cov.twist.twist;
+
+    const rclcpp::Time curr_time = rclcpp::Time(twist_with_cov.header.stamp);
+    if (curr_time > latest_odometry_.header.stamp) {
+      break;
+    }
+
+    const rclcpp::Duration time_diff = curr_time - prev_time;
+    const double time_diff_sec = time_diff.seconds();
+    if (time_diff_sec < 0.0) {
+      RCLCPP_WARN_STREAM(get_logger(), "time_diff_sec(" << time_diff_sec << ") < 0.0");
+      continue;
+    }
+
+    // quat to rpy
+    auto [ang_x, ang_y, ang_z] = quat_to_rpy(pose.orientation);
+
+    // rpy update
+    ang_x += twist.angular.x * time_diff_sec;
+    ang_y += twist.angular.y * time_diff_sec;
+    ang_z += twist.angular.z * time_diff_sec;
+    tf2::Quaternion quat;
+    quat.setRPY(ang_x, ang_y, ang_z);
+
+    // Convert twist to world frame
+    tf2::Vector3 linear_velocity(twist.linear.x, twist.linear.y, twist.linear.z);
+    linear_velocity = tf2::quatRotate(quat, linear_velocity);
+
+    // update
+    pose.position.x += linear_velocity.x() * time_diff_sec;
+    pose.position.y += linear_velocity.y() * time_diff_sec;
+    pose.position.z += linear_velocity.z() * time_diff_sec;
+    pose.orientation.x = quat.x();
+    pose.orientation.y = quat.y();
+    pose.orientation.z = quat.z();
+    pose.orientation.w = quat.w();
+    prev_time = curr_time;
+  }
+
+  // compare pose and latest_odometry_
+  const Pose latest_ekf_pose = latest_odometry_.pose.pose;
+  PoseStamped ekf_to_odom;
+  ekf_to_odom.pose = tier4_autoware_utils::inverseTransformPose(pose, latest_ekf_pose);
+  ekf_to_odom.header.stamp = latest_odometry_.header.stamp;
+  ekf_to_odom.header.frame_id = "base_link";
+  const geometry_msgs::msg::Point pos = ekf_to_odom.pose.position;
+  const auto [ang_x, ang_y, ang_z] = quat_to_rpy(ekf_to_odom.pose.orientation);
```

Comment:
It seems like the header information is not necessary, so I believe you can write simpler.
```suggestion
  Pose ekf_to_odom;
  ekf_to_odom = tier4_autoware_utils::inverseTransformPose(pose, latest_ekf_pose);
  const geometry_msgs::msg::Point pos = ekf_to_odom.position;
  const auto [ang_x, ang_y, ang_z] = quat_to_rpy(ekf_to_odom.orientation);
```

Reply:
That is correct. Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5439/commits/4b2ebe79c5fe063ea9f79a00b61c9157978a0612